### PR TITLE
Cloud Ingest (S3) and Structured File Import (csv, excel, parquet)

### PR DESCRIPTION
**Pull request by jharwig**
- [x] @joeferner
- [ ] @kunklejr @sfeng88 @diegogrz
- [ ] @mwizeman
- [ ] @joeybrk372 @rygim @jharwig @EvanOxfeld

Testing Instructions:
Cloud Ingest: File import -> Cloud Ingest -> S3. Connect with basic/session auth ingest one or more files. When finished, Open should select them.
Structured File: Test with csv, excel, parquet. Example parquet file in web/plugins/structured-ingest/parquet/src/test/resources

CHANGELOG
Added: Cloud Ingest support with new plugin for S3.
Added: Structured file ingest mapping interface. Ingest spreadsheets (csv, excel) and parquet files.

screens

### CODE REVIEW

File: web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/package.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "visallo-web-product-graph",
+  "description": "",
+  "main": "plugin.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (joeferner) [inline web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/package.json]: remove this line

File: web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/BrowserDirectoryItem.jsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                    <span className="name" title={name}>{name}</span>
+                    {type === 'file' ? (
+                        <span className="size">{F.bytes.pretty(size)}</span>
+                    ) : null}
+                </li>
+            );
+        },
+
+        onKeyUp(e) {
+            // TODO: handle enter
```
Comment (joeferner) [inline web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/BrowserDirectoryItem.jsx]: should we handle enter?

File: web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/CredentialsChooser.jsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            return {
+                selected: this.props.authenticationId || DEFAULT_AUTH_ID,
+                hideError: false
+            }
+        },
+
+        render() {
+            const { registry, errorMessage, loading } = this.props;
+            const { selected, hideError } = this.state;
+            const types = _.sortBy(registry[AUTH_EXTENSION_POINT], r => r.displayName.toLowerCase());
```
Comment (joeferner) [inline web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/CredentialsChooser.jsx]: should this be `toLocaleLowerCase`

File: web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/BasicAuth.jsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        },
+
+        getInitialState() {
+            return {
+                accessKey: '', secret: ''
+            }
+        },
+
+        componentDidMount() {
+            // FIXME: remove, just for faster dev testing
```
Comment (joeferner) [inline web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/BasicAuth.jsx]: remove

File: web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/BasicAuth.jsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                    <button
+                        disabled={!valid}
+                        onClick={this.connect}
+                        className={(loading ? "loading " : "") + "btn btn-primary"}>{i18n('connect')}</button>
+                    </div>
+                </div>
+            );
+        },
+
+        // FIXME: can't delete last letter
```
Comment (joeferner) [inline web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/BasicAuth.jsx]: is this still an issue?

File: web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/SessionAuth.jsx [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        },
+
+        getInitialState() {
+            return {
+                accessKey: '', secret: '', token: ''
+            }
+        },
+
+        componentDidMount() {
+            // FIXME: remove, just for faster dev testing
```
Comment (joeferner) [inline web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/SessionAuth.jsx]: remove

File: web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/SessionAuth.jsx [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                        <button
+                            disabled={!valid}
+                            onClick={this.connect}
+                            className={(loading ? "loading " : "") + "btn btn-primary"}>{i18n('connect')}</button>
+                    </div>
+                </div>
+            );
+        },
+
+        onKeyDown(event) {
```
Comment (joeferner) [inline web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/SessionAuth.jsx]: does this have a similar problem with deleting the last letter as above?

File: web/plugins/structured-ingest/core/src/main/java/org/visallo/web/structuredingest/core/routes/Ingest.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+
+        if (preview) {
+            return previewIngest(user, workspaceId, authorizations, optionsJson, publish, vertex, rawPropertyValue, parseMapping);
+        } else {
+            return enqueueIngest(user, workspaceId, authorizations, graphVertexId, mapping, optionsJson, publish);
+        }
+    }
+
+    private ClientApiObject enqueueIngest(User user, @ActiveWorkspaceId String workspaceId, Authorizations authorizations, @Required(name = "graphVertexId") String graphVertexId, @Required(name = "mapping") String mapping, @Optional(name = "parseOptions") String optionsJson, @Optional(name = "publish", defaultValue = "false") boolean publish) {
```
Comment (joeferner) [inline web/plugins/structured-ingest/core/src/main/java/org/visallo/web/structuredingest/core/routes/Ingest.java]: I'm guessing this code is from the previous CSV ingest but can you clean up the formatting of this or remove the annotations.

File: web/plugins/structured-ingest/core/src/main/java/org/visallo/web/structuredingest/core/routes/Ingest.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        StructuredIngestQueueItem queueItem = new StructuredIngestQueueItem(workspaceId, mapping, user.getUserId(), graphVertexId, StructuredIngestProcessWorker.TYPE, new ParseOptions(optionsJson), publish, authorizations);
+        this.longRunningProcessRepository.enqueue(queueItem.toJson(), user, authorizations);
+        return VisalloResponse.SUCCESS;
+    }
+
+    private ClientApiObject previewIngest(User user, @ActiveWorkspaceId String workspaceId, Authorizations authorizations, @Optional(name = "parseOptions") String optionsJson, @Optional(name = "publish", defaultValue = "false") boolean publish, Vertex vertex, StreamingPropertyValue rawPropertyValue, ParseMapping parseMapping) throws Exception {
```
Comment (joeferner) [inline web/plugins/structured-ingest/core/src/main/java/org/visallo/web/structuredingest/core/routes/Ingest.java]: same for this line

**Comment by joeferner** [comment]
I'm guessing you didn't want to rewrite CSV import in React :)

File: web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/BasicAuth.jsx (hunk 5 above)
Comment (jharwig) [inline web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/BasicAuth.jsx]: no, will remove

**Comment by jharwig** [comment]
@joeferner I did consider it, but the code was too large for the timeframe

File: web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/SessionAuth.jsx (hunk 7 above)
Comment (jharwig) [inline web/plugins/ingest-cloud-s3/src/main/resources/org/visallo/web/ingest/cloud/s3/js/auth/SessionAuth.jsx]: nope

**Pull request merged**